### PR TITLE
Virtual results

### PR DESCRIPTION
**Pull request by JensGM**
Introduce a new syntax for storing multiple results from a single
function.

```python
@xun.function()
def f(arg):
    yield even(arg) is arg * 2
    yield odd(arg) is arg * 2 - 1

@f.interface
def even(arg):
    yield from f(arg)

@f.interface
def odd(arg):
    yield from f(arg)
```

The `xun.Function.interface` defines a callable that is reliant on
another function to produce it's result. Which function is specified by
syntax `yield from <call-to-xun-function>`. For example, `yield from f(1)`
indicates that a call to `f(1)` is required to produce a result for the
interface.

### CODE REVIEW

File: xun/functions/blueprint.py [hunk 1]
```diff
@@ -59,17 +59,21 @@ def run(self, driver=None, store=None, client_store=None):
...
         function_images = {
-            name: func.callable(extra_globals={'_xun_store': store})
+            name: func.callable
             for name, func in self.functions.items()
         }
 
+        print("!!!!", function_images)
+        for name, func in function_images.items():
+            print('\t', name, ':', func.globals)
+
```
Comment (JensGM) [inline xun/functions/blueprint.py:76]: delete:
```suggestion
```

File: xun/functions/driver/driver.py [hunk 2]
```diff
@@ -14,8 +18,26 @@ def _exec(self, graph, entry_call, function_images, store_accessor):
...
     def __call__(self, graph, entry_call, function_images, store_accessor):
         return self.exec(graph, entry_call, function_images, store_accessor)
+
+    @staticmethod
+    def compute_and_store(callnode, func, store_accessor):
+        args, kwargs = store_accessor.resolve_call_args(callnode)
+        results = func(*args, **kwargs)
+        results.send(None)
+        results.send(store_accessor)
+        print(store_accessor.store_result)
```
Comment (JensGM) [inline xun/functions/driver/driver.py:34]: Left over print
```suggestion
```

File: xun/functions/function.py [hunk 3]
```diff
@@ -246,54 +166,183 @@ def graph(self, *args, **kwargs):
         nx.DiGraph
             The call graph for the call
         """
-        return self.createGraphBuilder()(*args, **kwargs)
+        print('graph: ', self.name, args, kwargs)
+        print(self.code.graph)
```
Comment (JensGM) [inline xun/functions/function.py:170]: prints
```suggestion
```

File: xun/functions/transformations.py [hunk 4]
```diff
@@ -541,3 +553,110 @@ def visit_Call(self, node):
...
+#
+
+@pass_by_value
+def separate_interface_and_target(func_desc: FunctionDescription, target):
+    fdef = func_desc.ast.body[0]
+    interface_call = call_from_function_definition(fdef)
+
+    import astunparse
+    print(astunparse.dump(fdef))
+
```
Comment (JensGM) [inline xun/functions/transformations.py:569]: Left over print
```suggestion
```

File: xun/tests/reference/fibonacci.py [hunk 5]
```diff
@@ -3,6 +3,9 @@
 
 @xun.function()
 def fibonacci_number(n):
+    print(globals()['fibonacci_number'])
+    print('Hey', f_n_1, fibonacci_number)
+    print('ya', f_n_2)
```
Comment (JensGM) [inline xun/tests/reference/fibonacci.py:8]:
```suggestion
```

File: xun/tests/test_transformations.py [hunk 6]
```diff
@@ -181,30 +182,86 @@ def dummy():
...
+    def reference_source(arg):
+        import networkx as _xun_nx
+        _xun_graph = _xun_nx.DiGraph()
+        _xun_graph.add_edge(f(arg * 2), g(arg))
+        return _xun_graph
+
+    reference = reference_source.body[0].body
+
+    print(generated)
+
```
Comment (JensGM) [inline xun/tests/test_transformations.py:231]:
```suggestion
```

File: xun/tests/test_transformations.py [hunk 7]
```diff
@@ -181,30 +182,86 @@ def dummy():
...
+    def reference_source(arg):
+        from xun.functions import XunInterfaceError as _xun_InterfaceError
+        raise _xun_InterfaceError(
+            f'{f(arg * 2)} did not produce a result for {g(arg)}'
+        )
+
+    reference = reference_source.body[0].body
+
+    print(generated)
+
```
Comment (JensGM) [inline xun/tests/test_transformations.py:260]:
```suggestion
```

**Review by mrSundvoll — APPROVED** [review]

**Pull request merged**